pkgs/by-name/gm/gmt/package.nix

```
@@ -0,0 +1,106 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  cmake,
+  curl,
+  darwin,
+  fftwSinglePrec,
+  netcdf,
+  pcre,
+  gdal,
+  blas,
+  lapack,
+  glibc,
+  ghostscript,
+  dcw-gmt,
+  gshhg-gmt,
+  libxml2,
+}:
+stdenv.mkDerivation (finalAttrs: {
+  pname = "gmt";
+  version = "6.5.0";
+
+  src = fetchFromGitHub {
+    owner = "GenericMappingTools";
+    repo = "gmt";
+    rev = "refs/tags/${finalAttrs.version}";
+    hash = "sha256-KKIYhljCtk9t9CuvTLsSGvUkUwazWTm9ymBB3wLwSoI=";
+  };
```

@Bot-wxt1221 This fixes the build issue on `aarch64-darwin` for me. Not sure if it is the best way to address this. Maybe you have some helpful insight?
```suggestion
  };
  
  postPatch = lib.optionalString stdenv.isDarwin ''
    substituteInPlace src/gmt_vector.c \
      --replace-fail '__LAPACK_int' 'int'
  '';
```

---

I traced this down.

## Definition of `__LAPACK_int`

The actual definition on a vanilla Darwin system is in `/Library/Developer/CommandLineTools/SDKs/MacOSX${darwin.apple_sdk.MacOSX-SDK.version}.sdk/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/Headers/lapack_types.h`:

```c
// For platforms that are neither ILP64 nor LP64, this will default to LP64 (for instance, arm64_32).
#if defined( ACCELERATE_LAPACK_ILP64 )
    typedef long __LAPACK_int;
    typedef long __LAPACK_bool;
#else
    typedef int  __LAPACK_int;
    typedef int  __LAPACK_bool;  // Because the fortran logical is 4 bytes
#endif
```

The exact issue that is this only appeared in SDK `13.3` and the current SDK version I'm seeing in nixpkgs is `11.0`:

```bash
$ nix eval --impure 'github:NixOS/nixpkgs/master#legacyPackages.aarch64-darwin.darwin.apple_sdk.MacOSX-SDK.version'
"11.0.0"
```

## Usage in `gmt`

The only usage of `__LAPACK_int` is in this block of `src/gmt_vector.c`:

```c
#ifdef ACCELERATE_NEW_LAPACK
	/* Must cast the arguemnts to avoid a tantrum from Apple */
	dgemm_ ((const char * _Nonnull)"t", (const char * _Nonnull)tr,
            (const __LAPACK_int * _Nonnull)&na, (const __LAPACK_int * _Nonnull)&nb, (const __LAPACK_int * _Nonnull)&nc,
            (const double * _Nonnull)&one, (const double * _Nullable) A, (const __LAPACK_int * _Nonnull)&nd,
            (const double * _Nullable) B, (const __LAPACK_int * _Nonnull)&ne,
            (const double * _Nonnull)&zero, (double * _Nullable)C, (const __LAPACK_int * _Nonnull)&nf);
#else
	dgemm_ ("t", tr, &na, &nb, &nc, &one, A, &nd, B, &ne, &zero, C, &nf);
#endif
```

... and `ACCELERATE_NEW_LAPACK` is defined in `src/CMakeLists.txt`:

```cmake
# Need to know if macOS and kernel is >= 22.4.0 for newLapack syntax
if (CMAKE_HOST_SYSTEM_NAME STREQUAL "Darwin")
	execute_process (COMMAND bash "-c" "echo ${CMAKE_HOST_SYSTEM_VERSION} | awk -F. '{if ($1 > 22 || ($1 == 22 && $2 > 3)) {print 1} else {print 0}}'"
		OUTPUT_STRIP_TRAILING_WHITESPACE OUTPUT_VARIABLE ACCELERATE_NEW_LAPACK)
else (CMAKE_HOST_SYSTEM_NAME STREQUAL "Darwin")
	set (ACCELERATE_NEW_LAPACK 0)
endif (CMAKE_HOST_SYSTEM_NAME STREQUAL "Darwin")
```

Ah-ha! The build is happening on a kernel ≥13.3, but the Accelerate headers visible to the compiler at build-time are from SDK 11.0.

There is only one other place in the codebase where `ACCELERATE_NEW_LAPACK` is used:

```c
#ifndef ACCELERATE_NEW_LAPACK
/* This BLAS function is updated on macos 22.4 and higher but still included in the Accelerate Framework header.
 * Otherwise we declare it as an extern function here.  dsyev_ is not in BLAS so remains the same so far. */

extern int dgemm_ (char* tra, char* trb, int* na, int* nb, int* nc, double* alpha, double* a, int *nd, double* b, int *ne, double* beta, double* c, int* nf);

#endif
```

## How to fix

So, there are a few valid fixes, in order of invasiveness and, proportionally, correctness:

1. Keep the existing fix replacing `__LAPACK_int` with `int`.

    From `extern int dgemm_` we can see that the arguments are clearly `int*`.
    
    AFAICT will work, but it's not pretty.
    
2. Redefine `ACCELERATE_NEW_LAPACK` by prepending the following before the first use of `ACCELERATE_NEW_LAPACK`:

    ```c
    #undef ACCELERATE_NEW_LAPACK
    #define ACCELERATE_NEW_LAPACK 0
    ```
    
3. Make 2) above a patch and apply it only if building with an SDK older than 13.3:

    ```nix
    lib.versionOlder darwin.apple_sdk.MacOSX-SDK.version "13.3"
    ```
    
## Conclusion

This should give you what you need. If you run into issues @-mention me here so I see a notification 🙏